src/BaGet.Core/Content/DatabasePackageContentService.cs

```diff
@@ -0,0 +1,97 @@
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using BaGet.Core.Metadata;
+using BaGet.Core.Mirror;
+using BaGet.Core.Storage;
+using BaGet.Protocol;
+using NuGet.Versioning;
+
+namespace BaGet.Core.Content
+{
+    /// <summary>
+    /// Implements the NuGet Package Content resource. Supports read-through caching.
+    /// Tracks state in a database (<see cref="IPackageService"/>) and stores packages
+    /// using <see cref="IPackageStorageService"/>.
+    /// </summary>
+    public class DatabasePackageContentService : IBaGetPackageContentService
+    {
+        private readonly IMirrorService _mirror;
+        private readonly IPackageService _packages;
+        private readonly IPackageStorageService _storage;
+
+        public DatabasePackageContentService(IMirrorService mirror, IPackageService packages, IPackageStorageService storage)
+        {
+            _mirror = mirror ?? throw new ArgumentNullException(nameof(mirror));
+            _packages = packages ?? throw new ArgumentNullException(nameof(packages));
+            _storage = storage ?? throw new ArgumentNullException(nameof(storage));
+        }
+
+        public async Task<PackageVersionsResponse> GetPackageVersionsOrNullAsync(string id, CancellationToken cancellationToken = default)
+        {
+            // First, attempt to find all package versions using the upstream source.
+            var versions = await _mirror.FindPackageVersionsOrNullAsync(id, cancellationToken);
+
+            if (versions == null)
+            {
+                // Fallback to the local packages if the package couldn't be found
```

Comment:
TODO: This comment is confusing. `_mirror.FindPackageVersionsOrNullAsync` merges local + upstream sources but returns `null` if mirroring is disabled.